### PR TITLE
Script helpers

### PR DESCRIPTION
**Pull request by brownjava**
Noticed that I was writing a bunch of boilerplate code in my migration and query scripts, so I factored some of it out.  The main stuff is in ScriptHelpers.  There's runScript and runPipes (which just calls runScript).

The runScript function just sets up a working Config environment which is roughly the same as the normal Config, but includes two changes: (1) stage and dry run are passed on the command line instead of via environment variables, and (2) I've implemented a DataSource which provides a raw MySQL read-only connection to the DB.

The runPipes function does everything that runScript does, but allows you to express your script in terms of:

(1) A setup function which is run first
(2) A set of streams which will all be piped into each other
(3) A teardown function which is run after the terminal stream finishes

I've also added some helper functions to create read/write/transform pipes using async/await methods instead of callbacks, since the stream APIs all use callbacks and I find that annoying.

### CODE REVIEW

File: src/utils/PipeHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      let value;
+      let error;
+      try {
+        value = await handler();
+      } catch (e) {
+        error = e;
+        console.error(`Error reading from stream: ${error}`);
+      }
+      if (value) {
+        this.push(value);
```
Comment (timothyk7) [inline src/utils/PipeHelpers.ts]: what is this pushing into?

File: src/utils/PipeHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  new stream.Readable({
+    objectMode: true,
+    async read() {
+      let value;
+      let error;
+      try {
+        value = await handler();
+      } catch (e) {
+        error = e;
+        console.error(`Error reading from stream: ${error}`);
```
[the remark this answers is not in the document]
Comment (brownjava) [reply src/utils/PipeHelpers.ts]: Yeah...I think this code got copied from some callback code.  But you're right, storing the error is unnecessary.

File: src/utils/PipeHelpers.ts (hunk 1 above)
Comment (brownjava) [reply src/utils/PipeHelpers.ts]: This is just the streams API for read streams.  Read is called on you, and you respond with this.push.
https://nodejs.org/api/stream.html#stream_readable_streams

File: src/utils/ScriptHelpers.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  if (_.toLower(argDryRun) === 'false') {
+    dryRun = true;
+  }
+
+  // Let script validate args
+  const validateArgsResult = argValidator(opt.argv);
+  if (validateArgsResult) {
+    // tslint:disable-next-line no-console
+    console.error(`Invalid arguments to script: ${validateArgsResult}\n`
+      + `usage: ${process.argv[0]} ${process.argv[1]} -s|--stage stage [--d|--dryRun dryRun] ${argDescription}`);
```
[the remark this answers is not in the document]
Comment (brownjava) [reply src/utils/ScriptHelpers.ts]: Typically [] denotes optionality.  In this case, stage is essentially required for the script to work.

It looks like I actually made this default to int, but we should probably just bail if it's not provided.

[the remark this answers is not in the document]
Comment (brownjava) [reply src/utils/ScriptHelpers.ts]: Fixed.

```
> node src/scripts/GatherOrgInfo.js
Please specify a stage
usage: /Users/jeremy/.nvm/versions/node/v8.9.3/bin/node /Volumes/classy/projects/wepay_tokens/src/scripts/GatherOrgInfo.js -s|--stage stage [--d|--dryRun dryRun] [PayPal|Auth]
```

File: src/utils/ScriptHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    require('../DataSources/Clients'),
+    new class extends DataSource {
+      private port?: number;
+      private user?: string;
+      private password?: string;
+
+      public async initialize(config: DataSourceConfig): Promise<void> {
+        let port: number|undefined;
+        switch (await config.get('stage')) {
+          case 'prod': port = 9306; break;
```
[the remark this answers is not in the document]
Comment (brownjava) [reply src/utils/ScriptHelpers.ts]: The class is actually anonymous (save for name() which is required to be implemented by the abstract parent, but is essentially never used).  If we ever add other DBs (which I will likely do at some point) we can either implement them here, or change this name if we implement them with some other data source.

File: src/utils/ScriptHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  argDescription: string = '',
+  argValidator: ValidateArgsFunction = args => undefined): void => {
+
+  // Parse args
+  const opt = Getopt.create([
+    ['s', 'stage=STAGE'],
+    ['d', 'dryRun=DRYRUN'],
+  ]).parseSystem();
+
+  const argStage: string = <string> _.get(opt.options, 'stage', 'int');
```
[the remark this answers is not in the document]
Comment (brownjava) [reply src/utils/ScriptHelpers.ts]: You can find these in the lodash definition files.

```
        get<TObject extends object, TKey extends keyof TObject, TDefault>(
            object: TObject | null | undefined,
            path: TKey | [TKey],
            defaultValue: TDefault
        ): TObject[TKey] | TDefault;
```

So in this case (un-genericized) it's basically any.

[the remark this answers is not in the document]
Comment (brownjava) [reply src/utils/ScriptHelpers.ts]: You can do this "better" with generics but in this case the cost is just too high. It's easier to explicitly cast at the end.

**Pull request merged**